authz/rbac_translator.go

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright 2021 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Package authz exposes methods to manage authorization within gRPC.
+package authz
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	v3routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	v3matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+)
+
+type header struct {
+	Key    string
+	Values []string
+}
+
+type peer struct {
+	Principals []string
+}
+
+type request struct {
+	Paths   []string
+	Headers []header
+}
+
+type rule struct {
+	Name    string
+	Source  peer
+	Request request
+}
+
+// Represents the SDK authorization policy provided by user.
+type authorizationPolicy struct {
+	Name       string
+	DenyRules  []rule `json:"deny_rules"`
+	AllowRules []rule `json:"allow_rules"`
+}
+
+func principalOr(principals []*v3rbacpb.Principal) *v3rbacpb.Principal {
+	return &v3rbacpb.Principal{
+		Identifier: &v3rbacpb.Principal_OrIds{
+			OrIds: &v3rbacpb.Principal_Set{
+				Ids: principals,
+			},
+		},
+	}
+}
+
+func principalAnd(principals []*v3rbacpb.Principal) *v3rbacpb.Principal {
+	return &v3rbacpb.Principal{
+		Identifier: &v3rbacpb.Principal_AndIds{
+			AndIds: &v3rbacpb.Principal_Set{
+				Ids: principals,
+			},
+		},
+	}
+}
+
+func permissionOr(permission []*v3rbacpb.Permission) *v3rbacpb.Permission {
+	return &v3rbacpb.Permission{
+		Rule: &v3rbacpb.Permission_OrRules{
+			OrRules: &v3rbacpb.Permission_Set{
+				Rules: permission,
+			},
+		},
+	}
+}
+
+func permissionAnd(permission []*v3rbacpb.Permission) *v3rbacpb.Permission {
+	return &v3rbacpb.Permission{
+		Rule: &v3rbacpb.Permission_AndRules{
+			AndRules: &v3rbacpb.Permission_Set{
+				Rules: permission,
+			},
+		},
+	}
+}
+
+func getStringMatcher(value string) *v3matcherpb.StringMatcher {
+	if value == "*" {
```

Comment:
Oh another option here, to use more Go-isms:

```go
switch {
	case value == "*":
		return ...
	case strings.HasSuffix(value, "*"):
		return ...
	case strings.HasPrefix(value, "*"):
		return ...
	default: // (or end switch here, either way)
		return ... // exact
}
```

Optional; it's functionally the same, but it would look a little nicer this way IMO.  Same for the `HeaderMatcher` version.